design-proposals/vm-backend-storage.md

```diff
@@ -0,0 +1,82 @@
+# Overview
+Today, anything that happens on the host side of VMIs (i.e. virt-launcher) is transient.  
+Any change a VMI makes outside of its own file-system is lost after a shutdown.  
+Such changes include EFI firmware settings and information stored inside the TPM.  
+This design provides a way to persist small VM-specific files used by virt-launcher.
+
+## Motivation
+Though we now support exposing an emulated TPM device to guests (vTPM), using the device in any meaningful way would require persisting data across reboots.  
+As we do not currently have a way to do that, all data stored by the VMI into the vTPM is lost after a shutdown (a soft reboot keeps the same virt-launcher, avoiding losses).
+
+## Goals
+- Being able to persist modifications of various emulated devices across reboots for a given VM
+- Enable live migration of VMs that use backend storage
+- Avoid consuming large amounts of storage
+- Prevent virt-launcher for VM X from seeing data that belongs to VM Y
+
+## Non Goals
+- Persisting everything. Some information that would usually persist on a non-ephemeral host just don't matter to us.
+- Supporting user modifications of the backend storage. Users whishing to inject their own files should do so at their own risk.
+
+## Definition of Users
+This feature is intended for end-users who wish to preserve the state of some emulated device across reboots.
+
+## User Stories
+As a VM user, I want to be able to store the encryption key of my VMs partitions inside the vTPM (example: BitLocker)
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+The storage used for this would be 1 PVC per VM, and would be created on first start of a VM that needs backend storage.
+The PVC would be RWX to ensure a VM can be started on a node, then stopped and re-started on another node.
+The PVC would be added to the virt-launcher compute container as a volume, but never assigned to the user's VM.  
+The PVC would be mounted in a directory at the root of the filesystem. Appropriate symbolic links will be created before starting the libvirt domain.
+The PVC would be owned-by the VM, and therefore will get deleted when its associated VM is deleted.
+
+For now, the only 2 candidates features for backend storage are:
+- The persistent TPM state, which takes about 6KiB when freshly created. We have yet to find its maximum size.
+- The OVMF vars, which can take up to 528KiB
+I propose finding the biggest size these can take, multiply the sum by 2 and round up to the nearest MiB to allow for future use-cases.
+The total will likely add up to 2MiB, 3MiB tops.
+
+## API Examples
+Currently, the TPM device API doesn't have any option.  
+We will add a `persistent` boolean option that will default to false.  
```

Comment:
I'm wondering whether the default should be false.
Isn't it a bit confusing that by default the TPM is non-functional? Perhaps it will cause workloads to fail and force the user to find this spec field.

Reply:
Good point, but I'm not sure how to handle upgrades from previous versions then... Would we add "persistent: false" to all existing VM(I)s on upgrade?

Reply:
Sounds good to me :)

Reply:
I actually disagree about defaulting to true. Enabling persistence requires setting up a RWX FS storage class and adding it to the kubevirt CR. Then the extra PVC costs storage and time, especially on migrations.
In other words, you probably don't want TPM persistence enabled unless you really need it.
Defaulting to false would also save a lot of headaches to people who already wrote yaml files for their VM(I)s and expect them to behave the same after upgrade...

Reply:
Alright, I'm convinced :)
Just one last thought - maybe we should at least warn about vTPM not being persistent? My concern is that something inside the guest would assume it has a valid TPM therefore things would break after reboot. The VM owner might get confused, IOW it wouldn't be obvious why things broke. Maybe we can add another condition, or event? WDYT?